File: csharp/perfect-numbers/PerfectNumbers.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
        => number <= 0 ? throw new ArgumentOutOfRangeException(nameof(number))
        : number.AliquotSum() == number ? Classification.Perfect
        : number.AliquotSum() > number ? Classification.Abundant
        : Classification.Deficient;


    public static int AliquotSum(this int number) 
        => Enumerable.Range(1, number / 2).Where(n => number % n == 0).Sum();
}

File: fsharp/perfect-numbers/PerfectNumbers.fs
module PerfectNumbers

type Classification =
    | Perfect
    | Abundant
    | Deficient


let classify n : Classification option =
    if n < 1 then
        None
    else
        let aliquotSum = [ 1 .. (n / 2) ] |> Seq.filter (fun i -> n % i = 0) |> Seq.sum

        match compare aliquotSum n with
        | 0 -> Perfect
        | 1 -> Abundant
        | _ -> Deficient
        |> Some
